Add: [Actions] release workflow for both releases and nightlies

TrueBrain opened the pull request:
Fixes #8330
Fixes #8309

Depends on https://github.com/OpenTTD/workflows/pull/10

A nightly run looks something like this:
https://github.com/TrueBrain/OpenTTD/actions/runs/420435699

A release run something like this:
https://github.com/TrueBrain/OpenTTD/actions/runs/420414947
(I tagged a fictional version in my own fork; you can ignore that completely)

----

## Targets currently not included

- ARM version of Mac OS. This is basically done and ready to ship, but we will wait till after this PR.
- ARM version of Windows. This requires a few more patches before it compiles successful. The rest of the workflow is prepared for this.
- Emscripten. This requires review of a few PRs, but otherwise this workflow is prepared to accept that too.
- AppImage. There is a PR to introduces most basics of AppImage to the repo; it needs integration with the release workflow in such way we minimize diskspace and still have a workable image.

## Choices made

- NSIS is only running for releases, not for nightlies.
- `.deb` builds are only made for releases, not for nightlies.
- Universal MacOS builds will be a PR further down the line, if it turns out we want them / our user-base wants them.
- We might attach our binaries to GitHub Releases for releases, but not right now. This will be an additional PR.

TrueBrain committed:
Fix: [bundle] custom launch.sh is no longer required for Mac OS

Co-authored-by: Owen Rudge <owen@owenrudge.net>

TrueBrain committed:
Fix: [bundle] postfix the architecture for Mac OS bundles

This in preparation for other architectures, like arm64.

TrueBrain committed:
Fix: [bundle] postfix the distribution for Linux bundles

TrueBrain committed:
Add: [Actions] release workflow for both releases and nightlies

This has several ways of being triggered:
- When creating a new release via the GitHub interface. Fully
  automated that will produce new binaries, upload them, and it
  will even update the website to tell about the new version.
- When triggered in an automated way from OpenTTD/workflows to
  start a nightly.
- Manually via the Release workflow, which accepts branches,
  Pull Requests and tags to build.

Rerunning a job is safe and should be without issues. Everything
retriggers and updates what-ever might have been broken. In fact,
except for dates, it should produce identical results.

Co-authored-by: Charles Pigott <charlespigott@googlemail.com>

glx22 commented:
I'll look at `OPTION_TOOLS_ONLY`

LordAro commented:
My particular answers for To Consider:
* Do we need -dbg.deb?
   Yes - matches having PDB on Windows, and good for consistency
* Do we want Debian builds? (we now only do Ubuntu)
   Yes
* Do we want NSIS runs for every nightly, or only for releases (like it is now).
   IMO no
* Do we want .deb builds for every nightly, or only for releases (like it is now).
   IMO no
* Do we want Mac OS universal builds?
   Yes
* Should we attach the binaries to GitHub Releases too? And if so, including the checksums? And if so, all three (MD5, SHA-1, SHA-256), or only SHA-256?
   IMO no to all. No point mixing where people get downloads from.

orudge commented:
I'll cover the macOS ARM builds and Universal builds, but will likely wait for this PR to be merged first.

michicc commented:
Attaching binaries to GitHub for releases is probably useful as there will be people that look for them on the GitHub Release tab first. I don't think having nightlies here as well makes sense, will just clutter the list way too much.

TrueBrain commented:
> 
> 
> Attaching binaries to GitHub for releases is probably useful as there will be people that look for them on the GitHub Release tab first. I don't think having nightlies here as well makes sense, will just clutter the list way too much.

I tempt to agree with you. And yes, I meant for releases only :D Oops .. should have been more clear :P For nightlies it is not even possible.
But for now I will not implement it; we can add it if we like it in a later PR :)

Look at me, scoping the PR :D W00p!

TrueBrain commented:
Unsure if `.github/changelog.sh` is a good location for that script ... but I also didn't find a better one. Suggestions are welcome!

TrueBrain committed:
Fix: [CMake] automatically detect dependencies for .deb packages

TrueBrain commented:
Fixed `.deb` dependencies, and postponed `-dbg.deb` (as it seems CPack doesn't support this currently for applications; only for "components"). So this is now ready for review!

TrueBrain committed:
Remove: [Azure Pipelines] So Long, and Thanks for All the Fish

Azure Pipelines has build our releases for two years now, but we
are finally switching to GitHub Actions. This to bring the full
workflow to a single place, making it easier for people to see
what is going on and how to influence the process.

glx22 commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,590 @@
+            echo "Internal error: branch name is empty."
+            echo "An earlier step should have prevented this from happening."
+            echo "Cancelling build, as without a branch name we cannot store the artifacts"
+          fi
```
I fail to see how and where the build is cancelled.

TrueBrain replied:
That is not a you problem, but a me problem :) Now it does cancel :D Nicely spotted :D

LordAro commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,590 @@
+name: Release
+
+on:
+  workflow_dispatch:
+    inputs:
+      ref:
+        description: 'Ref to build (for Pull Requests, use refs/pull/NNN/head)'
+        required: true
+  repository_dispatch:
+    # client_payload should be the same as the inputs for workflow_dispatch.
+    types:
+    - Build*
```
I don't quite understand what these are for, can you explain?
Relatedly, why are the nightlies not triggered with a cron, e.g. https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#onschedule ?

TrueBrain replied:
Like I don't know about `schedule` .. pfft :P See https://github.com/OpenTTD/workflows/pull/10 :)

`workflow_dispatch` means we have a new button in the GitHub interface that allows us to make a binary release of anything, anytime. This makes rebuilding easier, but also building PRs possible.

`repository_dispatch` is similar, but for automated triggers. A while ago we decided to move all those workflows into https://github.com/OpenTTD/workflows . This has mostly to do with forks. If you fork this repo, these days by default schedules are disabled; but this was not always the case. Additionally, if you enable it once for what-ever reason, it remains on for 90 days. So that is pretty noisy. So we decided to move all the schedule stuff to https://github.com/OpenTTD/workflows, as it is very unlikely anyone would fork that.

Additionally, the schedule code is not straightforward. It does not support timezones, but also it cannot check if there was anything new. That complexity is now all in https://github.com/OpenTTD/workflows/pull/10, as it really has nothing to do with OpenTTD.

In https://github.com/OpenTTD/workflows you will find more lovely glue-code we use for our infra :) It also triggers eints commits, publishes the docs, rebuilds the website, etc :)

LordAro commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,590 @@
+      folder: ${{ steps.metadata.outputs.folder }}
+
+    steps:
+    - name: Checkout (Release)
+      if: github.event_name == 'release'
+      uses: actions/checkout@v2
+      with:
+        # We generate a changelog; for this we need the full git log.
+        fetch-depth: 0
+
+    - name: Checkout (Manual)
+      if: github.event_name == 'workflow_dispatch'
+      uses: actions/checkout@v2
+      with:
+        ref: ${{ github.event.inputs.ref }}
+        # We generate a changelog; for this we need the full git log.
+        fetch-depth: 0
+
+    - name: Checkout (Trigger)
+      if: github.event_name == 'repository_dispatch'
+      uses: actions/checkout@v2
+      with:
+        ref: ${{ github.event.client_payload.ref }}
+        # We generate a changelog; for this we need the full git log.
+        fetch-depth: 0
```
Would it be possible to set a 'ref' variable to use rather than duplicating this 3 times?

TrueBrain replied:
Yes, kind of. But it would require a `bash` rule with the `if` logic in there to deduce which ref to use. The downside of that is, it is really easy to fuck that up. This now is trivially correct, and is why I prefer it. It does require a step extra .. but as you can see in a step later where we figure out the branch-name .. that code is not pretty, and really easy to mess up :) (which I also did, as it turns out :P)

So if you don't mind, I prefer this way. If you do mind, I can rewrite it .. but no guarantees :D

(I hope GitHub comes with some updates soon that I can also remove the other bash blob, and integrate it into these blobs .. but the `actions/checkout` needs some changes for that :D).

LordAro commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,590 @@
+        FOLDER_BRANCHES: openttd-branches
+
+    - name: Remove VCS information
+      run: find . -iname .hg -or -iname .git -or -iname .svn | xargs rm -rf
```
Do we really need to bother with .hg or .svn ? It's also only ever going to be '.git' (lowercase) in the repo root, so `rm -rf .git` ought to suffice?

TrueBrain replied:
I just copy/pasted from Azure workflow, but sure, I will clean this up :)

LordAro reviewed (commented):
More questions than problems

LordAro reviewed (approved):
About time we had working nightlies again

glx22 reviewed (approved):
Ok for me too.
